src/messages.cpp

```diff
@@ -238,159 +241,235 @@ bool Messages::has_undisplayed_messages()
+     * The 'for (;;)' block below is nearly completely method-agnostic, save for the `player_messages.impl_->history(i)` call.
+     *
+     * In case of NEWEST-TOP, the 'i' variable easily enough corresponds to the newest message.
+     * In case of OLDEST-TOP, the 'i' variable must be flipped- meaning the highest value of 'i' returns the result for the lowest value of 'i', etc.
+     * To achieve this, the 'flip_message' variable is set to either the value of 'msg_count', or '0'. This variable is then subtracted from 'i' in each call to player_messages.impl_->history();
+     *
+     * 'offset' refers to the corresponding message that will be displayed at the very TOP of the message box window.
+     *  NEWEST-TOP: 'offset' starts simply at '0' - the very top of the window.
+     *  OLDEST-TOP: 'offset' is set to the maximum value it could possibly be. That is- 'msg_count-bottom'. This way, the screen starts with the scrollbar all the way down.
+     * 'retrieve_history' refers to the line that should be displayed- this is either 'i' if it's NEWEST-TOP, or a flipped version of 'i' if it's OLDEST-TOP.
+     */
+    int offset = log_from_top ? 0 : ( msg_count - bottom );
+    const int flip = log_from_top ? 0 : msg_count - 1;
+    int retrieve_history;
+
+    for( ;; ) {
+        werase( w );
+        draw_border( w );
+        mvwprintz( w, bottom + 1, 32, c_red, _("Press %s to return"), ctxt.get_desc("QUIT").c_str() );
+        draw_scrollbar( w, offset, bottom, msg_count, 1, 0, c_white, true );
 
         int line = 1;
         int lasttime = -1;
         for( int i = offset; i < msg_count; ++i ) {
-            if (line > bottom) {
+            retrieve_history = abs( i - flip );
+            if( line > bottom ) {
                 break;
+                // This statement makes it so that no non-existent messages are printed (which usually results in a segfault)
+            } else if( retrieve_history >= msg_count ) {
+                continue;
             }
 
-            const game_message &m     = player_messages.impl_->history(i);
-            const nc_color col        = msgtype_to_color( m.type, false );
+            const game_message &m     = player_messages.impl_->history( retrieve_history );
             const calendar timepassed = calendar::turn - m.timestamp_in_turns;
-
-            if (timepassed.get_turn() > lasttime) {
-                mvwprintz(w, line++, 3, c_ltblue, _("%s ago:"),
-                    timepassed.textify_period().c_str());
+            const char *long_ago      = timepassed.textify_period().c_str();
+            nc_color col              = msgtype_to_color( m.type, false );
+
+            // Here we seperate the unit and amount from one another so that they can be properly padded when they're drawn on the screen.
+            // Note that the very first character of 'unit' is often a space (except for languages where the time unit directly follows the number.)
+            char unit[0];
```

Comment:
My compiler (MinGW32, -O3) gives the following disassembly:
```asm
338                 sscanf( long_ago, "%d%s", &amount, unit );
   0x0046006c <Messages::display_messages()+1308>:      lea    -0x120(%ebp),%eax
   0x00460072 <Messages::display_messages()+1314>:      movl   $0xcc87de,0x4(%esp)
   0x0046007a <Messages::display_messages()+1322>:      mov    %esi,(%esp)
   0x0046007d <Messages::display_messages()+1325>:      mov    %eax,0xc(%esp)
   0x00460081 <Messages::display_messages()+1329>:      lea    -0x108(%ebp),%eax
   0x00460087 <Messages::display_messages()+1335>:      mov    %eax,0x8(%esp)
   0x0046008b <Messages::display_messages()+1339>:      call   0xa6f368 <sscanf>
```
Which I think is `(%esp) == %esi == long_ago`, `0x4(%esp) == $0xcc87de == "%d%s"`, `0x8(%esp) == -0x108(%ebp) == &amount`, `0xc(%esp) == -0x120(%ebp) == unit`.

So `unit` actually has `0x120 - 0x108 == 0x18 == 24` byte space. But we can't rely on this behavior. For example, in [this disassembly](https://godbolt.org/g/vhZkFF), `unit` is passed to `sscanf` by `movq %rbp, %rcx`, and `%rbp` is the value of the stack pointer before calling `sscanf`. So `sscanf` will write beyond the stack pointer address and cause a stack overflow.

So it'll be better to explicitly give it a length, or even better, use some c++ facilities like std::istringstream so we won't have to care about overflow anymore.